backend/main_test.go

```diff
@@ -476,3 +465,80 @@ func TestImproveBadMethods(t *testing.T) {
 	postresp := makeHTTPRequest(postreq, handler, t)
 	checkResponseStatus(http.StatusMethodNotAllowed, postresp, t)
 }
+
+type failingAuthStub struct{}
+
+func (auth failingAuthStub) authenticate(next http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, "Auth failed", http.StatusUnauthorized)
+		return
+	}
+}
+
+func TestAuthMiddleware(t *testing.T) {
+	server := Server{store: in_memory_storage.MakeInMemStore()}
+	server.auth = failingAuthStub{}
+	handler := server.makeHandler()
+
+	testHandlerWithAuth := func(httpMethod, target string, body io.Reader) {
+		req := httptest.NewRequest(httpMethod, target, body)
+		resp := makeHTTPRequest(req, handler, t)
+		checkResponseStatus(http.StatusUnauthorized, resp, t)
+	}
+	teamID := "myteam"
+	periodID := "2019q1"
+
+	testHandlerWithAuth(http.MethodGet, "/api/team/"+teamID, nil)
+	testHandlerWithAuth(http.MethodGet, "/api/team/", nil)
+	testHandlerWithAuth(http.MethodPost, "/api/team/", nil)
+	testHandlerWithAuth(http.MethodPut, "/api/team/"+teamID, nil)
+	testHandlerWithAuth(http.MethodGet, "/api/period/"+teamID+"/"+periodID, nil)
+	testHandlerWithAuth(http.MethodGet, "/api/period/"+teamID+"/", nil)
+	testHandlerWithAuth(http.MethodPost, "/api/period/"+teamID+"/", nil)
+	testHandlerWithAuth(http.MethodPut, "/api/period/"+teamID+"/"+periodID, nil)
+
+	getreq := httptest.NewRequest(http.MethodGet, "/improve", nil)
```

Comment:
This could probably use a comment explaining that you're asserting here that `/improve` is not covered by authentication.

Either that or you could write another helper function called something like `assertNoAuthenticationFailure`, that makes the opposite assertion to the one above. You could either pass in the specific status code expected, or just assert that you get something in the `200` or `300` range.

Reply:
I added a comment